### PR TITLE
[REEF-476] Add IFileSystem

### PR DESCRIPTION
**Pull request by markusweimer**
This change adds the interface `Org.Apache.REEF.IO.FileSystem.IFileSystem`, a local implementation and tests.

JIRA:
  [REEF-476](https://issues.apache.org/jira/browse/REEF-476)

### CODE REVIEW

**Comment by markusweimer** [comment]
@beysims @shravanmn @dkm2110 can you please have a look and see whether this matches your needs?

**Comment by jwang98052** [comment]
I will look at it.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        private LocalFileSystem()
+        {
+        }
+
+        public Stream Open(Uri fileUri)
+        {
+            return File.Open(fileUri.LocalPath, FileMode.Open);
+        }
+
+        public Stream Create(Uri fileUri)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs]: LocalFileSystem  is more like a wrapper of System File and Directory. What's the point to wrap them?

File: lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.IO.FileSystem.Local
+{
+    /// <summary>
+    /// An implementation of IFileSystem on the filesystem of the host.
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs]: space : file system

**Comment by jwang98052** [comment]
I have completed my pass.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/Local/LocalFileSystem.cs]: The idea is similar to the Hadoop `FileSystem` class which abstracts the actual file system interface. That way, applications can be written that can run against the local file system or the HDFS one without change.

**Comment by jwang98052** [comment]
@beysims @shravanmn @dkm2110 can you please have a look and see whether this matches your needs before I merge it?

**Comment by jwang98052** [comment]
Shall I merge this one?

**Comment by markusweimer** [comment]
@jwang98052 Not yet. Let's give them till tomorrow morning to comment.

**Comment by markusweimer** [comment]
@jwang98052 The JIRA has been up for 2 weeks, and code has been available to look at for 10 days. I think it is safe to assume lazy consent on this. hence, go ahead and merge :-)

File: lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        /// <returns></returns>
+        /// <exception cref="IOException">If the URI couldn't be created.</exception>
+        Stream Create(Uri fileUri);
+
+        /// <summary>
+        /// Deletes the file under the given URI.
+        /// </summary>
+        /// <param name="fileUri"></param>
+        /// <exception cref="IOException"></exception>
+        void Delete(Uri fileUri);
```
Comment (shravanmn) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs]: Perhaps a return value indicating the status of delete

File: lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        /// <exception cref="IOException"></exception>
+        void CopyToLocal(Uri remoteFileUri, string localName);
+
+        /// <summary>
+        /// Copies the specified file to the remote location.
+        /// </summary>
+        /// <param name="localFileName"></param>
+        /// <param name="remoteFileUri"></param>
+        /// <exception cref="IOException"></exception>
+        void CopyFromLocal(string localFileName, Uri remoteFileUri);
```
Comment (shravanmn) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs]: can we not copy a directory?

File: lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        /// <summary>
+        /// Copies the file referenced  by sourceUri to destinationUri.
+        /// </summary>
+        /// <param name="sourceUri"></param>
+        /// <param name="destinationUri"></param>
+        /// <exception cref="IOException"></exception>
+        void Copy(Uri sourceUri, Uri destinationUri);
+
+        /// <summary>
+        /// Copies the remote file to a local file.
```
Comment (shravanmn) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/IFileSystem.cs]: Perhaps change the doc to directory if the functionality can, if not, any reason for not supporting directories?

**Pull request closed without merging**